---
title: "Single Peptide Results Analysis"
author: "Caleb Easterly"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

### Necessary packages:

```{r message=FALSE}
library(GO.db)
library(data.table)
library(dplyr)
library(stringr)
```

## Introduction

The single peptide analysis methods and results are contained in this document. 

### 20 Random peptides for Acetyl-CoA C-acetyltransferase activity
```{r}
go_target <- "GO:0003985" # acetyl-CoA C-acetyltransferase activity
twenty_peps <- read.table('data/acetyl-CoA_C-acetyltransferase_activity_twenty_peps.tabular',
                          stringsAsFactors=FALSE)[,1]
twenty_peps
```

```{r}
filterOntology <- function(go_list, ontology){
    if (ontology == "all") { return(go_list) }
    onts <- Ontology(go_list)
    return(go_list[onts == ontology & !is.na(onts)])
}
ontology <- "MF"
```

## Unipept

### Process

* Paste the tabular list of peptides into the Unipept ‘Metaproteomics Analysis’ web application (https://unipept.ugent.be/datasets)  
* Parameters:
    - Equate I and L: FALSE
    - Filter duplicate peptides: FALSE
    - Advanced missed cleavage handling: TRUE
* Download results
* Annotate each peptide with only the GO terms that are present in 5% or more of the proteins (percentages are returned in GO term column)

```{r}
add_coverage <- function(gos_col, go_protein_counts_col, total_protein_count_col){
  mapply(function(gos_col, go_protein_counts_col, total_protein_count_col){
    sapply(1:length(gos_col), function(i){
      gos_list <- strsplit(as.character(gos_col[i]), ',')[[1]]
      gos_protein_counts_list <- strsplit(as.character(go_protein_counts_col[i]), ',')[[1]]
      paste(mapply(function(gos, go_protein_counts) {
        coverage <- round(as.numeric(go_protein_counts) / total_protein_count_col[i], digits = 3)
        paste0(gos, '(', coverage, ')')
      }, gos=gos_list, go_protein_counts=gos_protein_counts_list),
      collapse=",")
    })
  }, gos_col=gos_col, go_protein_counts_col=go_protein_counts_col, total_protein_count_col)
}

filter_coverages <- function(go_lists){
  sapply(go_lists, function(go_list){
    go_list <- strsplit(go_list, ',')[[1]]
    #go_list
    coverages <- as.numeric(str_match(go_list, '\\((.+)\\)')[,2])
    paste(go_list[coverages > 0.05], collapse=',')
  })
} 

filter_go_mf <- function(go_list){
  go_list <- strsplit(as.character(go_list), ',')[[1]] 
  go_processed <- sapply(go_list, function(go){
    str_match(go, '(GO:\\d+)\\((.+)\\)')
  })
  go_processed[2,][Ontology(go_processed[2,]) == "MF"]
}

unipept_pep <- read.delim('data/unipept/737all-unipept-pept2go.tabular') %>%
  filter(str_detect(go_terms, go_target)) %>%
  rename(peptide = X.peptide) %>%
  mutate(go_coverage = add_coverage(go_terms, go_protein_counts, total_protein_count),
         filtered_gos = filter_coverages(go_coverage),
         go_mf_filtered = sapply(filtered_gos, function(go_list){
           paste(filter_go_mf(go_list), collapse=',')
         }))

uni <- unipept_pep %>%
  filter(peptide %in% twenty_peps)
uni <- uni[!duplicated(uni[c(1,3)]),] # remove duplicates
# unipept_hits_unfiltered <- as.character(unipept_pep_refiltered$go_terms)
# unipept_hits_unfiltered <- unique(unlist(strsplit(unipept_hits_unfiltered, ',')))
uni
# Term(setdiff(unipept_hits_unfiltered, unipept_hits))
```


<!-- Get the number of peptides (non-unique) at a peptide deformylase activity and spectral counts -->
<!-- ```{r} -->
<!-- uniWS[grepl("peptide deformylase activity", uniWS$uni_go_names),] -->
<!-- uniNS[grepl("peptide deformylase activity", uniNS$uni_go_names),] -->
<!-- ``` -->


## eggNOG mapper

* Use the Galaxy version of eggNOG mapper, on Galaxy-P
* Parameters:
    - eggNOG Database version: eggnog_4.5.1
    - Annotation type: DIAMOND
    - Scoring matrix and gap costs: PAM30, 9 and 1
    - Taxonomic Scope: Bacteria
    - Orthologs: use all orthologs
    - Gene Ontology evidence: use non-electronic terms
    - Seed Orthology Search Options
        - Min e-value: 200000
        - Min bit score: 20
* Download and compare GO terms
```{r}
peptidesNS <- read.delim("peptide_shaker_outputs/737NS_Peptide_Shaker_Peptide_Report.tabular",
                         stringsAsFactors = FALSE) %>%
    select(peptide = Sequence, countNS = "X.Validated.PSMs")

peptidesWS <- read.delim("peptide_shaker_outputs/737WS_Peptide_Shaker_Peptide_Report.tabular",
                         stringsAsFactors = FALSE) %>%
    select(peptide = Sequence, countWS = "X.Validated.PSMs") 

norm_ratio <- sum(peptidesWS$countWS) / sum(peptidesNS$countNS)
peptidesNS$countNS <- peptidesNS$countNS * sum(peptidesWS$countWS) / sum(peptidesNS$countNS)
peptides_all <- full_join(peptidesNS, peptidesWS, by = "peptide")
peptides_all[is.na(peptides_all)] <- 0
peptides_all
```

```{r}
eggnog <- read.delim("eggnog_results/diamond_annotations.tabular",
                     stringsAsFactors = FALSE,
                     header = FALSE) %>% 
    select(peptide = V1, protein = V2, gene = V5, go = V6, ko = V7, desc = V13) %>%
    filter(peptide %in% twenty_peps)
#eggnog
eggnog_w_counts <- left_join(eggnog,
                             peptides_all,
                             by = "peptide")

eggnog_w_counts$go <- sapply(eggnog_w_counts$go, function ( go ) {
      paste(filterOntology(str_split(go, ",")[[1]], ontology), collapse="; ")
    })
eggnog_w_counts <- eggnog_w_counts[eggnog_w_counts$go != "",]
em <- eggnog_w_counts %>% select(peptide, em_go = go)
em
```
Get the number of peptides (non-unique) at a peptide deformylase activity and spectral counts
```{r}
eggnog_w_counts$countNS
eggnog_w_counts$countWS
```

## BLASTP against UniProt


* Use the UniProtKB BLAST web search on each peptide, one-by-one
* Parameters
    - Target database: UniProtKB
    - E-Threshold: 10
    - Matrix: Auto
    - Filtering: None
    - Gapped: Yes
    - Hits: 50
* For each peptide, download the result list and get all GO terms and TaxID associated with that peptide
* To match Unipept, annotate each peptide with only the GO terms that are present in 5% or more of the proteins
* Get the most frequent protein name
* For taxonomy, we can also calculate the lowest common ancestor of each peptide (TODO)

```{r}
process_gos <- function(go_col){
  sapply(go_col, function(go_list){
    go_list <- strsplit(as.character(go_list), '; ')[[1]]
    go_list_mf <- go_list[str_detect(go_list, "F")]
    go_list_mf_formatted <- gsub("F:", "", go_list_mf)
    paste(go_list_mf_formatted, collapse="; ")
  })
}

blast <- read.delim('data/blast2go_outputs/uni_glucosyl_peps.txt') %>%
  select(peptide = SeqName,
         blast_go = GO.IDs) %>%
  mutate(blast_go = process_gos(blast_go))
blast <- blast[blast$blast_go != "",]
blast
```

```{r}
# blast_go <- rep("", length(peptides))
# blast_names <- rep("", length(peptides))
# files <- list.files('uniprot_blastp_outputs')
# for (i in 1:length(peptides)){
#   filename = paste('uniprot_blastp_outputs', paste(peptides[i], '.tab', sep=""), sep="/")
#   info = file.info(filename)
#   if (is.na(info$size)){
#     print(peptides[i])
#   }
#   if (length(rownames(info[info$size == 0, ])) != 1){
#     result <- read.delim(filename,
#                          stringsAsFactors = FALSE,
#                          na.strings = c("", "NA", "NaN"))
#     gos <- table(unlist(str_split(result$Gene.ontology..molecular.function., "; ")))/50
#     filtered_gos <- names(gos)[which(gos > 0.05)]
#     filtered_gos <- str_extract(filtered_gos, "GO:\\d+")
#     if(length(filtered_gos) == 0){ filtered_gos <- "" }
#     #str_extract(filtered_gos, "[\\w ]+\\w"))
#     filtered_gos <- filtered_gos[Ontology(filtered_gos) == ontology]
#     if(is.na(filtered_gos)) { filtered_gos <- ""}
#     blast_go[i] <- paste(filtered_gos, collapse = "; ")
#   }
# }
# blast <- data.frame(peptide = peptides, blast_go, stringsAsFactors = FALSE)
# blast$blast_names <- sapply(strsplit(blast$blast_go, "; "), function ( x ) { paste(Term(x), collapse="; ")})
# blast
```

```{r}
# format_to_fasta <- function(peptides){
#   unlist(sapply(peptides, function(peptide) {  list(paste0('>', peptide), peptide) }))
# }
# write.table(format_to_fasta(blast$peptide),
#                    file='data/peptide_deformylase_activity.fasta',
#             quote=FALSE, row.names=FALSE,)
```


Get the number of peptides (non-unique) at a peptide deformylase activity and spectral counts
```{r}
blast_w_counts <- left_join(blast,
                             peptides_all,
                             by = "peptide")
blast_deformyl <- blast_w_counts[grepl("peptide deformylase activity", blast_w_counts$blast_name),]
blast_deformyl_ws <- blast_deformyl[!is.na(blast_deformyl$countWS),]
blast_deformyl_ns <- blast_deformyl[!is.na(blast_deformyl$countNS),]
length(blast_deformyl$peptide)
sum(blast_deformyl_ws$countWS)
sum(blast_deformyl_ns$countNS)
```

## MetaGOmics

* Upload HOMD to metaGOmics
* Parameters:
    - Uniprot database: Uniprot sprot
    - Blast e-value cutoff: 1e-10
    - Use only top hit?: TRUE
* Result URL: https://www.yeastrc.org/metagomics/viewUploadedFasta.do?uid=MLRUUtSgjRbbO5td
* One-by-one, upload peptides and run
* Download results individually, combine into table

```{r}
twenty_peps_regex <-  paste(paste0('(',twenty_peps,')'), collapse='|')
                            
metagomics <- data.table(read.delim("data/metagomics/go_compare_624_625.txt", 
                         comment.char = "#") %>%
                           mutate(peptides = gsub(',$', '',
                                                  paste(Run.1.peptides,
                                                        Run.2.peptides, sep=","))) %>%
                         select(go = GO.acc, namespace = GO.aspect, name = GO.name,
                                peptides) %>% 
  filter(grepl(twenty_peps_regex, peptides)))
metagomics
# 
# Split peptides into individual rows
mg <- metagomics[ , list( peptides = unlist( strsplit( as.character(peptides) , ',' ) ) ) ,
                          by = list(go, namespace, name) ]
mg <- mg %>%
  group_by(peptides) %>%
  summarise(gos = paste0(go, collapse=",")) %>%
  filter(peptides %in% twenty_peps)
mg
```

```{r}
library(GO.db)
mg_go <- rep(0, length(peptides))
dir <- 'metaGOmics_single_peptides_outputs/'
files <- list.files(dir)
for (i in 1:length(peptides)){
  result <- read.delim(paste(dir, paste(peptides[i], '.txt', sep=""), sep=""),
                       stringsAsFactors = FALSE,
                       na.strings = c("", "NA", "NaN"),
                       comment.char = "#")
  if (length(result$GO.acc) > 0){
    gos <- filterOntology(result$GO.acc, ontology)
    gos <- gos[gos != ""]
  } else {
    gos <- ""
  }
  mg_go[i] <- paste(gos, collapse = "; ")
}
mg <- data.frame(peptide = peptides, mg_go, stringsAsFactors = FALSE)
mg$mg_names <- sapply(strsplit(mg$mg_go, "; "), function ( x ) { paste(Term(x), collapse="; ")})
mg
```
Get the number of peptides (non-unique) at a peptide deformylase activity and spectral counts
```{r}
metagomics <- read.delim("metaGOmics_results/go_compare_149_150.txt",
                             stringsAsFactors = FALSE,
                             comment.char = "#")
metagomics[metagomics$GO.name == "peptide deformylase activity",] %>%
  select(mg_NS = Run.1.PSM.count, mg_WS = Run.2.PSM.count)
length(mg[grepl("peptide deformylase activity", mg$mg_names),]$peptide)
```


## Combine all of the results:

```{r}
all_results <- plyr::join_all(list(blast, mg, uni, em), by = "peptide")
peptide <- "IFVCDD"
"BLAST"
all_results[grepl(peptide, all_results$peptide),][1,]$blast_names
"EM"
all_results[grepl(peptide, all_results$peptide),][1,]$em_names
"MG"
all_results[grepl(peptide, all_results$peptide),][1,]$mg_names
"Uni"
all_results[grepl(peptide, all_results$peptide),][1,]$uni_go_names
all_results
```

All of the results are below:

```{r results = 'asis'}
library(pander)
# knitr::kable(all_results)
pander::pander(all_results, split.cell = 80, split.table = Inf)
```


The question is, then, what the other terms are.

The other terms could be related to the 4 we found in several ways:

1) They could be more general terms (ancestors)  of the 4 we found
2) They could be more specific terms (descendants) of the 4 we found
3) They could be terms that are not ancestors or descendants. These may be 'extraneous' GO terms.

However, we can also consider that terms that are not ancestors or descendants can be either closely or distantly related to the 4 we found. For this analysis, let's define 'closely related' as 'an ancestor, descendant, or child of ancestor' of the 4 terms. Terms that are not closely related are declared to be extraneous. So, we have 5 categories:

1) original terms (BLAST terms)
2) ancestors
3) descendants
4) children of ancestors
5) extraneous

```{r}
library(httr)
library(jsonlite)
get_go_names <- function(id_vector){
  if (length(id_vector) > 1){
    base_url <- 'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/'
    terms <- str_replace(id_vector, ":", "%3A")
    joined_terms <- paste(terms, collapse="%2C")
    term_url <- paste(base_url, joined_terms, sep="")
    term_info <- GET(term_url, accept("application/json"))
    json <- toJSON(content(term_info))
    names <- unlist(fromJSON(json)$results$name)
    names
  }
}
```

Let's look at each of the latter 4 categories in turn (we already determined that 4 of each of the metaGOmics and eggNOG mapper terms are the 4 BLAST/Unipept terms).

### Ancestors
```{r}
# get all ancestors of the 4 terms
get_go_ancestors <- function(id_vector){
    base_url <- 'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/'
    terms <- str_replace(id_vector, ":", "%3A")
    joined_terms <- paste(terms, collapse="%2C")
    term_url <- paste(base_url, joined_terms, '/ancestors?relations=is_a', sep="")
    term_info <- GET(term_url, accept("application/json"))
    json <- toJSON(content(term_info))
    names <- unlist(fromJSON(json)$results$ancestors)
    names
}
get_go_descendants <- function(id_vector){
    base_url <- 'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/'
    terms <- str_replace(id_vector, ":", "%3A")
    joined_terms <- paste(terms, collapse="%2C")
    term_url <- paste(base_url, joined_terms, '/descendants?relations=is_a', sep="")
    term_info <- GET(term_url, accept("application/json"))
    json <- toJSON(content(term_info))
    names <- unlist(fromJSON(json)$results$descendants)
    names
}
get_children <- function(goids){
    base_url <- 'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/'
    terms <- str_replace(goids, ":", "%3A")
    joined_terms <- paste(terms, collapse="%2C")
    term_url <- paste(base_url, joined_terms, sep="")
    term_info <- GET(term_url, accept("application/json"))
    json <- toJSON(content(term_info))
    children <- fromJSON(json)$results$children
    children_is_a <- lapply(children, function(x) unlist(x[x$relation == "is_a", ]$id))
    return(children_is_a)
}
get_length <- function(target){
  if (is.na(target) || length(target) == 0 || is.null(target) || target == ""){
    return(0)
  } else {
    return(length(target))
  }
}
```

```{r}
# TODO: Check why unipept's split doesn't have a space
library(ggplot2)
n <- nrow(all_results)
output <- data.frame(Peptide = all_results[,'peptide'],
	                     BLAST = character(n),
	                     Unipept = character(n),
	                     Equal = logical(n),
	                     stringsAsFactors = FALSE)
#for (n in 1:1){
for (n in 1:nrow(all_results)){
  print(all_results[n, 'peptide'])
  blast <- unique(c(str_split(all_results[n, 'blast_go'], "; ", simplify=TRUE)))
  blast <- blast[blast != ""]
  print("blast terms:")
  print(get_go_names(blast))
  print("")
  
  uni <- unique(c(str_split(all_results[n, 'uni_go_mf'], ";", simplify=TRUE)))
  uni <- uni[uni != ""]
  print("unipept terms:")
  print(uni)
  print(get_go_names(uni))
  print("")
  
  eggnog_list <- unique(c(str_split(all_results[n, 'em_go'], "; ", simplify =TRUE)))
  metaGOmics_list <- unique(c(str_split(all_results[n, 'mg_go'], "; ", simplify=TRUE)))
  print("eggNOG terms")
  print(get_go_names(eggnog_list))
  print("metaGOmics terms")
  print(get_go_names(metaGOmics_list))
  print("")
  
  output$em.list[n] <- paste(eggnog_list, collapse="; ")
  output$mg.list[n] <- paste(metaGOmics_list, collapse="; ")
  
  ancestors <- setdiff(unlist(get_go_ancestors(blast)), blast)
  descendants <- setdiff(unlist(get_go_descendants(blast)), blast)
  children_of_ancestors <- ""
  if (length(ancestors) > 0){
    children_of_ancestors <- setdiff(unlist(get_children(matrix(ancestors))), ancestors)
  }
  closely_related <- union_all(blast, descendants, ancestors, children_of_ancestors)
  
  em_ancestors <- intersect(eggnog_list, ancestors)
  mg_ancestors <- intersect(metaGOmics_list, ancestors)
  uni_ancestors <- intersect(uni, ancestors)
  print("Ancestors:")
  print("eggnog:")
  print(get_go_names(em_ancestors))
  print("metagomics:")
  print(get_go_names(mg_ancestors))
  print("")
  
  em_descendants <- intersect(descendants, eggnog_list)
  mg_descendants <- intersect(descendants, metaGOmics_list)
  uni_descendants <- intersect(descendants, uni)
  em_ancestors_kids <- setdiff(unlist(intersect(eggnog_list, children_of_ancestors)), blast)
  mg_ancestors_kids <- setdiff(unlist(intersect(metaGOmics_list, children_of_ancestors)), blast)
  uni_ancestors_kids <- setdiff(unlist(intersect(uni, children_of_ancestors)), blast)
  print("Descendants:")
  print("eggnog:")
  print(get_go_names(em_descendants))
  print("metagomics:")
  print(get_go_names(mg_descendants))
  print("")
  print("Ancestors' kids:")
  print("eggnog:")
  print(get_go_names(em_ancestors_kids))
  print("metagomics:")
  print(get_go_names(mg_ancestors_kids))
  print(mg_ancestors_kids)
  print("")
  
  em_extraneous <- setdiff(eggnog_list, closely_related)
  mg_extraneous <- setdiff(metaGOmics_list, closely_related)
  uni_extraneous <- setdiff(uni, closely_related)
  print("Extraneous:")
  print("eggnog:")
  print(get_go_names(em_extraneous))
  print(em_extraneous)
  print(length(em_extraneous))
  print("metagomics:")
  print(get_go_names(mg_extraneous))
  print("")
  
  
  em_blast <- intersect(eggnog_list, blast)
  mg_blast <- intersect(metaGOmics_list, blast)
  uni_blast <- intersect(uni, blast)
  
  term_df <- data.frame(
    typeOfTerm = rep(c("BLAST", "Descendants", "Ancestors", "Ancestors' kids", "Extraneous"), 4),
    NumTerms = c(get_length(blast), 0, 0, 0, 0,
                 get_length(em_blast), get_length(em_descendants), get_length(em_ancestors), get_length(em_ancestors_kids), get_length(em_extraneous),
                 get_length(mg_blast), get_length(mg_descendants), get_length(mg_ancestors), get_length(mg_ancestors_kids), get_length(mg_extraneous),
                 get_length(uni_blast), get_length(uni_descendants), get_length(uni_ancestors), get_length(uni_ancestors_kids), get_length(uni_extraneous)),
    Tool = rep(c("BLAST2GO - NCBI (nr)", "EggNOG mapper", "MetaGOmics", "Unipept"), each = 5)
  )
  term_df$typeOfTerm <- factor(term_df$typeOfTerm, 
                               levels = c("Extraneous", "Ancestors' kids", "Ancestors", "Descendants", "BLAST"))
  
  
  plot <- ggplot(term_df, aes(x = Tool, y = NumTerms, fill = typeOfTerm)) +
    geom_bar(color = "black", stat = "identity", position = "stack", size = 0.1) +
    scale_fill_manual("Type of term", values = c("Extraneous" = "#D576EE",
                                                 "Ancestors' kids" = "#A4A233", "Ancestors" = "#E57D74",
                                                 "Descendants" = "#4FAFE4", "BLAST" = "#5DBA85")) +
    ylab("Number of Terms") +
    # ylab(" ") + xlab(" ") +
    scale_y_continuous(breaks = round(seq(0, max(sapply(unique(term_df$typeOfTerm), function( tool ) {
                                            sum(term_df[term_df$typeOfTerm == tool,]$NumTerms)  
                                          })), by = 2) + 2,1)) +
    ggtitle(all_results[n, 'peptide']) +
    #theme_update(plot.title = element_text(hjust = 0.5, size = 6)) +
    theme_classic() + 
    theme(legend.title=element_text(size=22),
          #legend.title=element_blank(),
          plot.title = element_text(size = 8, hjust = 0.5, face = "bold"),
          legend.text=element_text(size=18),
          legend.position="right",
          #legend.position="none",
          legend.spacing.x = unit(0.8, "cm"),
          axis.text.x = element_text(size=8, angle = 90, hjust = 0.95, vjust = 0.2),
          axis.title.x = element_text(margin = margin(t = 10, r = 0, b = 0, l = 0)))
  print(plot)
  ggsave(paste("plots/peptide_deformylase/", all_results[n, 'peptide'], ".png", sep=""), plot = plot, width = 4, height = 4, dpi = 800, units = "in")
}
#pander::pander(output, split.cell = 10, split.table = Inf)
``` 
```{r}
length(closely_related)
#!grepl("GO:\\d{7}",closely_related)
```

```{r}
#str_split(output$em.paths[1], ' ', simplify = TRUE)
#shortest_path("GO:0003723", "GO:0065007")
#get_go_ancestors("GO:0065007")
#get_go_names(get_go_ancestors("GO:0008150"))
get_go_names(str_split(output$BLAST[1], '; ', simplify = TRUE))
```


```{r}
# Check if there is no overlap between categories
n <- 1
length(str_split(output$BLAST[n],'; ', simplify=TRUE)) + length(str_split(output$em.descendants[n],'; ', simplify=TRUE)) +
length(str_split(output$em.ancestors[n],'; ', simplify=TRUE)) + length(str_split(output$em.ancestors.kids[n],'; ', simplify=TRUE)) +
length(str_split(output$em.extraneous[n],'; ', simplify=TRUE))
length(eggnog_list)
```


```{r}
#shortest_path("GO:0006417",)
#get_go_ancestors("GO:0003723")
#get_go_ancestors("GO:0003735")
#get_go_ancestors("GO:0005840")
#ancestors <- get_go_ancestors("GO:0006412")
#ancestors <- ancestors[1:2]
#ancestors
#children_of_ancestors <- setdiff(unlist(get_children(ancestors)), ancestors)
#children_of_ancestors

# AFLPGSLVDTRPVR
# head(get_go_names('GO:0065007'))
# head(get_go_names('GO:0006412'))
# head(get_go_names('GO:0003723'))
# head(get_go_names('GO:0003735'))
# head(get_go_names('GO:0005840'))

# DIAMQIAAVNPTYLNREEVPTEVIEHEK
# head(get_go_names('GO:0005618'))
# head(get_go_names('GO:0030312'))
```


```{r}
#row <- all_results[all_results$peptide == "IVGAKGKAALLR",]
row <- all_results[all_results$peptide == "IFVCDDGNGVVR",]
row$blast_names
row$em_names
row$mg_names
row$uni_go_names
```


<!-- Calculate proportions to answer three questions: -->
<!-- 1) How many of Uniprot's terms does the tool pick up? -->
<!-- 2) What is the proportion of total terms from the tool that are extraneous? -->

<!-- #### Eggnog -->
<!-- ```{r} -->
<!-- # answer to 1 -->
<!-- length(intersect(eggnog_list, buni_split)) / length(buni_split) -->

<!-- # answer to 2 -->
<!-- length(em_diff_with_kids)/length(eggnog_list) -->
<!-- ``` -->

<!-- #### MetaGOmics -->
<!-- ```{r} -->
<!-- # answer to 1 -->
<!-- length(intersect(metaGOmics_list, buni_split)) / length(buni_split) -->

<!-- # answer to 2 -->
<!-- length(mg_diff_with_kids)/length(metaGOmics_list) -->
<!-- ``` -->


<!-- metaGOmics -->
<!-- ```{r} -->
<!-- diff <- setdiff(metaGOmics_list, full_tree) -->
<!-- get_go_names(diff) -->
<!-- ``` -->

<!-- eggNOG mapper -->
<!-- ```{r} -->
<!-- em_diff <- setdiff(eggnog_list, full_tree) -->
<!-- get_go_names(em_diff) -->
<!-- ``` -->

<!-- Visualize the overlap between the full Blast+Unipept tree (descendants, ancestors) and the eggNOG and metaGOmics term lists. -->
<!-- ```{r fig.width=4, fig.height=4} -->
<!-- library(VennDiagram) -->
<!-- grid.newpage() -->
<!-- grid.draw(venn.diagram( -->
<!--     list("eggnog" = eggnog_list, "blast+unipept" = full_tree, "metagomics" = metaGOmics_list), -->
<!--     NULL)) -->
<!-- file.remove(list.files(pattern = "VennDiagram.*log")) # venn diagram log files -->
<!-- ``` -->

<!-- ## Future directions -->
<!-- 1) repeat this for the other 4 peptides -->
<!-- 2) how do we handle terms that are not descendants or ancestors? We could define some distance cutoff, and say that everything beyond that is a false hit. For example, we could say that if the shortest path between a metaGOmics or eggNOG term and any term in the full B+U tree has length greater than or equal to 2 than it is a false hit. -->

<!-- ### Get distance -->
<!-- ```{r} -->
<!-- get_paths <- function(from, to){ -->
<!--     base_url <- 'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/' -->
<!--     from_clean <- str_replace(from, ":", "%3A") -->
<!--     to_clean <- str_replace(to, ":", "%3A") -->
<!--     term_url <- paste(base_url, from_clean, -->
<!--                       "/paths/", to_clean, "?relations=is_a", sep="") -->
<!--     paths <- GET(term_url, accept("application/json")) -->
<!--     json <- toJSON(content(paths)) -->
<!--     names <- fromJSON(json)$results -->
<!--     names -->
<!-- } -->

<!-- shortest_path <- function(go1, go2){ -->
<!--     paths1_2 <- get_paths(go1, go2) -->
<!--     paths2_1 <- get_paths(go2, go1) -->
<!--     paths <- c(paths1_2, paths2_1) -->
<!--     min(sapply(paths, length)) -->
<!-- } -->

<!-- shortest_path("GO:1901136", "GO:0008150") -->
<!-- ``` -->

<!-- ## GO glossary -->

<!-- Here, I get the names of all the above GO terms. -->

<!-- ``` -->
<!-- library(httr) -->
<!-- library(jsonlite) -->
<!-- base_url <- 'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/' -->
<!-- term_url <- paste(base_url, 'GO%3A0008150%2CGO%3A0008152', sep="") -->
<!-- term_info <- GET(term_url, verbose(), accept("application/json")) -->
<!-- json <- toJSON(content(term_info)) -->
<!-- df <- fromJSON(json)$results -->
<!-- ``` -->